### PR TITLE
Arrow keys control left eye.

### PR DESCRIPTION
**Pull request by rictic**
Screen Shot 2020-09-25 at 5 03 45 PM

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -9,51 +9,79 @@ use wasm_bindgen::JsCast;
...
-    if !up {
-        return Some(true);
-    }
+pub struct Game {
+    x: f64,
+    y: f64,
+}
+
+#[wasm_bindgen]
+impl Game {
```
Comment (cdata) [inline src/lib.rs:19]: Ooooo `impl`

File: src/lib.rs [hunk 2]
```diff
@@ -9,51 +9,79 @@ use wasm_bindgen::JsCast;
...
+    // These params describe whether the player is currently holding down each
+    // of the buttons.
+    pub fn draw(&mut self, up: bool, down: bool, left: bool, right: bool) -> Option<bool> {
+        let document = web_sys::window().unwrap().document()?;
```
Comment (cdata) [inline src/lib.rs]: Why can you use `?` on `document()` but not on `window()`?

File: src/lib.rs [hunk 3]
```diff
@@ -9,51 +9,79 @@ use wasm_bindgen::JsCast;
...
+        let canvas = document.get_element_by_id("canvas")?;
+        let canvas: web_sys::HtmlCanvasElement =
+            canvas.dyn_into::<web_sys::HtmlCanvasElement>().ok()?;
+
+        let context = canvas
+            .get_context("2d")
+            .ok()??
```
Comment (cdata) [inline src/lib.rs:30]: How does `??` work? Is it a double use of the `?` operator, or a distinct operator?

File: src/lib.rs [hunk 4]
```diff
@@ -9,51 +9,79 @@ use wasm_bindgen::JsCast;
...
+            .dyn_into::<web_sys::CanvasRenderingContext2d>()
+            .ok()?;
+
+        context.clear_rect(0.0, 0.0, canvas.width().into(), canvas.height().into());
 
-    context.begin_path();
+        if up {
+            self.y -= 1.0
```
Comment (cdata) [inline src/lib.rs:37]: This is why we need a physics engine :v: :v: :v:

File: src/lib.rs [hunk 5]
```diff
@@ -9,51 +9,79 @@ use wasm_bindgen::JsCast;
...
-    Some(true)
+        context.stroke();
+
+        Some(true)
+    }
+}
+
+#[wasm_bindgen]
+pub fn make_game() -> Game {
+    Game { x: 0.0, y: 0.0 }
```
Comment (cdata) [inline src/lib.rs:86]: :joystick:

File: www/index.js [hunk 6]
```diff
@@ -46,14 +46,16 @@ document.body.appendChild(output);
...
+  const result = game.draw(
       curInput.up, curInput.down, curInput.left, curInput.right);
   if (result == null) {
     output.innerText = 'Failed to draw!';
   } else {
-    output.innerText = 'All is well. (hold up for smile)';
+    output.innerText = 'All is well.';
   }
-  requestAnimationFrame(main);
+  requestAnimationFrame(drawOneFrame);
```
Comment (cdata) [inline www/index.js:59]: Is this the only way to re-enter WASM? Is there no way to do this within Rust?

**Review by cdata — APPROVED** [review]

File: src/lib.rs (hunk 4 above)
Comment (rictic) [reply src/lib.rs:37]: :rocket:

File: src/lib.rs (hunk 2 above)
Comment (rictic) [reply src/lib.rs]: Because I missed that one!

Done

File: src/lib.rs (hunk 3 above)
Comment (rictic) [reply src/lib.rs:30]: Double use of the `?`

File: www/index.js (hunk 6 above)
Comment (rictic) [reply www/index.js:59]: We'd need the ability to send a closure from Rust to the JS bindings to register the animation frame. It's doable, just another thing to learn

File: src/lib.rs (hunk 1 above)
Comment (rictic) [reply src/lib.rs:19]: methods! kinda

on a struct!

**Pull request merged**